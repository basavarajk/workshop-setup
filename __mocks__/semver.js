module.exports = {
  satisfies: jest.fn(),
}
