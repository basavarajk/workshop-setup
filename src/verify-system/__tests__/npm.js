import cpMock from 'child_process'
import {satisfies as satisfiesMock} from 'semver'
import oneLine from '../../utils/one-line'
import validateNpm from '../npm'

jest.mock('child_process')

test('returns npm version mismatch', () => {
  const mockNpmVersion = '3.11.3'
  setup({
    mockNpmVersion,
    satisfiesNpm: false,
    mockYarnVersion: () => {
      throw new Error('blah')
    },
  })
  const desiredNpm = '^4.0.3'
  const result = validateNpm(desiredNpm)()
  expect(result).toContain(desiredNpm)
  expect(result).toContain(mockNpmVersion)
  expect(result).toMatchSnapshot()
})

function setup({mockNpmVersion = '4.0.3', satisfiesNpm = true} = {}) {
  satisfiesMock.mockClear()
  cpMock.execSync.mockClear()
  satisfiesMock.mockImplementation(satisfiesMockImplementation)
  cpMock.execSync.mockImplementation(execSyncMock)

  function satisfiesMockImplementation(actual, desired) {
    if (actual === mockNpmVersion) {
      return satisfiesNpm
    } else {
      console.error(
        oneLine`
          called satisifies with "${desired}, ${actual}".
          ${actual} does not match npm@${mockNpmVersion}
        `,
      )
      throw new Error('satisfies mock implementation insufficient')
    }
  }

  function execSyncMock(command) {
    if (command.includes('npm')) {
      return callAndReturn(mockNpmVersion)
    } else {
      console.error(
        oneLine`
          called execSync with "${command}".
          This does not include npm
        `,
      )
      throw new Error('execSync mock implementation insufficient')
    }
  }

  function callAndReturn(thing) {
    if (typeof thing === 'function') {
      thing = thing()
    }
    return thing
  }
}
