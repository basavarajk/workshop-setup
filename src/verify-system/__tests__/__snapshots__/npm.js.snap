// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns npm version mismatch 1`] = `"The installed version of npm (3.11.3) does not satisfy the desired range of ^4.0.3. Please at least have a version within the specified version range. You can install the latest version by running \`npm install --global npm@^4.0.3\`."`;
